sympy/core/cache.py

```diff
@@ -70,7 +70,9 @@ def func_wrapper(func):
         def wrapper(*args, **kwargs):
             try:
                 retval = cfunc(*args, **kwargs)
-            except TypeError:
+            except TypeError as e:
+                if not e.args or not e.args[0].startswith('unhashable type:'):
+                    raise
```

Comment:
Parsing error strings seems flakey to me.

What if the cached function fails because of an unhashable type?

Reply:
The more correct way would be to check `hash(args)`, but this is more expensive, so probably not a good idea for this function.